File: components/sound/src/main/java/org/storm/sound/manager/SoundManager.java
package org.storm.sound.manager;

import org.storm.sound.Sound;
import org.storm.sound.exception.SoundException;

import java.util.HashMap;
import java.util.Map;

/**
 * A SoundManager is helpful way of tracking and using sounds within a game
*/
public class SoundManager {

  private static final String SOUND_NOT_FOUND_ERROR_TEMPLATE = "no sound found for id: %s";

  private final Map<String, Sound> soundMap;

  public SoundManager() {
    this.soundMap = new HashMap<>();
  }

  /**
   * Adds the given sound to the SoundManager under the given id
   *
   * @param id the id reference for the Sound
   * @param sound Sound to track
   */
  public void add(String id, Sound sound) {
    this.soundMap.put(id, sound);
  }

  /**
   * Removes the sound associated to the given id reference
   *
   * @param id the id reference for the Sound
   */
  public void remove(String id) {
    this.soundMap.remove(id);
  }

  /**
   * Plays the Sound associated with the given id
   *
   * @param id the id reference for the Sound
   */
  public void play(String id) {
    if (!this.soundMap.containsKey(id)) {
      throw new SoundException(String.format(SOUND_NOT_FOUND_ERROR_TEMPLATE, id));
    }
    this.soundMap.get(id).play();
  }

  /**
   * Stops the Sound associated with the given id
   *
   * @param id the id reference for the Sound
   */
  public void stop(String id) {
    if (!this.soundMap.containsKey(id)) {
      throw new SoundException(String.format(SOUND_NOT_FOUND_ERROR_TEMPLATE, id));
    }
    this.soundMap.get(id).stop();
  }

  /**
   * Pauses the Sound associated with the given id
   *
   * @param id the id reference for the Sound
   */
  public void pause(String id) {
    if (!this.soundMap.containsKey(id)) {
      throw new SoundException(String.format(SOUND_NOT_FOUND_ERROR_TEMPLATE, id));
    }
    this.soundMap.get(id).pause();
  }

  /**
   * Adjusts the volume of the Sound associated with the given id
   *
   * @param id the id reference for the Sound
   * @param volume the new volume for the Sound
   */
  public void adjustVolume(String id, double volume) {
    if (!this.soundMap.containsKey(id)) {
      throw new SoundException(String.format(SOUND_NOT_FOUND_ERROR_TEMPLATE, id));
    }
    this.soundMap.get(id).adjustVolume(volume);
  }

  /**
   * Adjusts the volume of all the Sounds tracked by the SoundManager
   *
   * @param volume the new volume for all the Sounds
   */
  public void adjustAllVolume(double volume) {
    this.soundMap.forEach((id, sound) -> sound.adjustVolume(volume));
  }

  /**
   * Stops all sounds held by the manager
   */
  public void stopAll() {
    this.soundMap.forEach((id, sound) -> sound.stop());
  }

  /**
   * Pauses all sounds held by the manager
   */
  public void pauseAll() {
    this.soundMap.forEach((id, sound) -> sound.pause());
  }
}

File: components/sound/src/main/kotlin/org/storm/sound/manager/SoundManager.kt
package org.storm.sound.manager

import com.fasterxml.jackson.core.type.TypeReference
import org.storm.core.asset.AssetManager
import org.storm.core.context.Context
import org.storm.core.utils.observation.Observable
import org.storm.core.utils.observation.Observer
import org.storm.sound.Sound
import org.storm.sound.context.BGM_VOLUME
import org.storm.sound.context.EFFECT_VOLUME
import org.storm.sound.context.MASTER_VOLUME
import org.storm.sound.context.VOICE_VOLUME
import org.storm.sound.exception.SoundException

/**
 * A SoundManager is a helpful way of tracking and using sounds within a game
 */
class SoundManager(
    private val assetSourceId: String = "local-storage",
    private val assetManager: AssetManager,
    private val sounds: MutableMap<String, Sound> = mutableMapOf(),
): Observer {

    fun loadSound(id: String) {
        val sound = assetManager.getAsset("sound", id, assetSourceId, object: TypeReference<Sound>() {})
        add(id, sound)
    }

    /**
     * Adds the given sound to the SoundManager under the given id
     *
     * @param id the id reference for the Sound
     * @param sound Sound to track
     */
    fun add(id: String, sound: Sound) {
        sounds[id] = sound
    }

    /**
     * Removes the sound associated to the given id reference
     *
     * @param id the id reference for the Sound
     */
    fun remove(id: String) {
        sounds.remove(id)
    }

    /**
     * Plays the Sound associated with the given id
     *
     * @param id the id reference for the Sound
     */
    fun play(id: String) {
        sounds[id] ?: throw SoundException("no sound found for id: $id")
        sounds[id]!!.play()
    }

    /**
     * Stops the Sound associated with the given id
     *
     * @param id the id reference for the Sound
     */
    fun stop(id: String) {
        sounds[id] ?: throw SoundException("no sound found for id: $id")
        sounds[id]!!.stop()
    }

    /**
     * Pauses the Sound associated with the given id
     *
     * @param id the id reference for the Sound
     */
    fun pause(id: String) {
        sounds[id] ?: throw SoundException("no sound found for id: $id")
        sounds[id]!!.pause()
    }

    /**
     * Adjusts the volume of the Sound associated with the given id
     *
     * @param id the id reference for the Sound
     * @param volume the new volume for the Sound
     */
    fun adjustVolume(id: String, volume: Double) {
        sounds[id] ?: throw SoundException("no sound found for id: $id")
        sounds[id]!!.adjustVolume(volume)
    }

    /**
     * Adjusts the volume of all the Sounds tracked by the SoundManager
     *
     * @param volume the new volume for all the Sounds
     */
    fun adjustAllVolume(volume: Double, type: String? = null) {
        sounds.filter {
            type == null || it.value.soundType == type
        }.forEach { (_, sound: Sound) ->
            sound.adjustVolume(volume)
        }
    }

    /**
     * Stops all sounds held by the manager
     */
    fun stopAll() {
        sounds.forEach { (_, sound: Sound) -> sound.stop() }
    }

    /**
     * Pauses all sounds held by the manager
     */
    fun pauseAll() {
        sounds.forEach { (_, sound: Sound) -> sound.pause() }
    }

    override fun update(o: Observable) {
        if (o !is Context) return

        adjustAllVolume(o.BGM_VOLUME * o.MASTER_VOLUME, Sound.Type.BGM.value)
        adjustAllVolume(o.EFFECT_VOLUME * o.MASTER_VOLUME, Sound.Type.EFFECT.value)
        adjustAllVolume(o.VOICE_VOLUME * o.MASTER_VOLUME, Sound.Type.VOICE.value)
    }
}
